// @workspace/db/src/schema/materials.ts
import { pgTable, text, timestamp, integer, pgEnum } from "drizzle-orm/pg-core";
import { createInsertSchema } from "drizzle-zod";
import { z } from "zod/v4";

export const materialAccessEnum = pgEnum("material_access", ["public", "restricted", "confidential"]);
export const materialStatusEnum = pgEnum("material_status", ["draft", "published", "archived"]);

export const materialsTable = pgTable("materials", {
  id: text("id").primaryKey(),
  materialId: text("material_id").notNull().unique(),
  title: text("title").notNull(),
  altTitle: text("alt_title"),
  creator: text("creator"),
  description: text("description"),
  date: text("date"),
  categoryId: text("category_id"),
  access: materialAccessEnum("access").notNull().default("public"),
  format: text("format"),
  fileSize: text("file_size"),
  pages: integer("pages"),
  language: text("language"),
  publisher: text("publisher"),
  contributor: text("contributor"),
  subject: text("subject"),
  type: text("type"),
  source: text("source"),
  rights: text("rights"),
  relation: text("relation"),
  coverage: text("coverage"),
  identifier: text("identifier"),
  // ISAD(G) fields
  archivalHistory: text("archival_history"),
  custodialHistory: text("custodial_history"),
  accessionNo: text("accession_no"),
  scopeContent: text("scope_content"),
  arrangement: text("arrangement"),
  // Technical / Preservation
  sha256: text("sha256"),
  scanner: text("scanner"),
  resolution: text("resolution"),
  physicalLocation: text("physical_location"),
  physicalCondition: text("physical_condition"),
  bindingType: text("binding_type"),
  cataloger: text("cataloger"),
  dateCataloged: text("date_cataloged"),
  // OAIS
  sipId: text("sip_id"),
  aipId: text("aip_id"),
  ingestDate: text("ingest_date"),
  ingestBy: text("ingest_by"),
  fixityStatus: text("fixity_status"),
  // Other
  preferredCitation: text("preferred_citation"),
  fileUrl: text("file_url"),
  thumbnailUrl: text("thumbnail_url"),
  status: materialStatusEnum("status").notNull().default("published"),
  createdBy: text("created_by"),
  createdAt: timestamp("created_at").notNull().defaultNow(),
  updatedAt: timestamp("updated_at").notNull().defaultNow(),
});

export const insertMaterialSchema = createInsertSchema(materialsTable).omit({ id: true, materialId: true, createdAt: true, updatedAt: true });
export type InsertMaterial = z.infer<typeof insertMaterialSchema>;
export type Material = typeof materialsTable.$inferSelect;
